Migrate to new Repository

DomFleischmann opened the pull request.

Commit:
Initial commit

Signed-off-by: Dominik Fleischmann <dominik.fleischmann@canonical.com>

Commit:
Hard Code only Python Module

Signed-off-by: Dominik Fleischmann <dominik.fleischmann@canonical.com>

Commit:
Merge remote-tracking branch 'old-repo/main' into main

johnsca commented on k8s_service.py:
```diff
@@ -0,0 +1,69 @@
+        self._relation_name = relation_name
+
+        self.framework.observe(
+            charm.on[relation_name].relation_joined, self._on_relation_joined
```
I was initially going to recommend switching this to `relation_created` since you really only need to do it once for each relation, but then I remembered that I mentioned in the charm review that the port value should be passed in somehow rather than this class prescribing that it come from a config option (maybe the workload doesn't support a configurable port) or what the config option name is (maybe the workload has multiple ports to manage).  That made me think about what would be needed to handle changes to the value as well and I realized that `relation_created` isn't sufficient even with the value being passed in, since if it changes, you'll need to update existing relations as well.

So I think it would actually be cleaner in the end to have the port be passed in to the constructor and then just iterate over all current relations here in `__init__` and set the values for each one. That way, all relations will always get the latest value, and Juju will automatically handle the case where the data is the same anyway and not trigger new hooks, and you don't need to worry about observing any events here, nor does the charm author need to worry about observing `config_changed`.  I feel like this is probably going to be a good general pattern for relations where one side is just publishing data without expecting any sort of response, which is quite a few of them.

johnsca replied:
So, something like:

```python
class ProvideK8sService(Object):
    def __init__(self, charm, relation_name, port):
        super().__init__(charm, relation_name)
        for relation in self.model.relations[relation_name]:
            relation.data[charm.app].update({
                "service-name": charm.app.name,
                "service-port": str(port),
            })
```

DomFleischmann replied:
I really like this approach! Less moving parts.

johnsca commented on k8s_service.py:
```diff
@@ -0,0 +1,69 @@
+
+    @property
+    def service_port(self):
+        return self.state.service_port
```
Similar to the above comment, I think it would actually be simpler in the end to just read the data in `__init__` and set attributes and not have to worry about any stored state, events, or anything.  So, it would just become this:

```python
class RequireK8sService(Object):
    def __init__(self, charm, relation_name):
        super().__init__(charm, relation_name)
        relation = self.model.get_relation(relation_name)
        data = relation.data[relation.app] if relation else {}
        self.service_name = data.get("service-name")
        self.service_port = data.get("service-port")
        self.is_available = (self.service_name and self.service_port)
```

This does raise the question of whether this interface should support multiple relations or not. If not, `model.get_relation()` can raise a `TooManyRelatedAppsError` exception if `limit` is not set on the relation in the `metadata.yaml` and more than one relation is established. But I could also certainly see a charm using this to provide some sort of ingress for multiple services, so maybe it would be worth supporting it; that would mean this would need to iterate over all relations and expose a list of service name / port pairs instead of a single pair, of course.

johnsca replied:
Oh, hrm.  This doesn't include the `k8s_service_available` event, which could be useful for the charm if the relation using this is non-optional, or for detecting changed data. That's going to need stored state for at least the hash of the values to tell if they've changed. So maybe something like this, instead, (and I'll add in the multiple relation support, too, in case you want that):

```python
class RequireK8sService(Object):
    state = StoredState()
    on = RequireK8sServiceEvents()

    def __init__(self, charm, relation_name):
        super().__init__(charm, relation_name)
        self.state.set_default(data_hash=None)

        self.services = []
        for relation in sorted(self.model.relations[relation_name], key=attrgetter('id')):
            service_name = relation.data[relation.app].get("service-name")
            service_port = relation.data[relation.app].get("service-port")
            if service_name and service_port:
                self.services.append((service_name, service_port))
        self.is_available = bool(self.services)

        data_hash = md5(str(self.services).encode('utf8')).hexdigest()
        if self.is_available and self.state.data_hash != data_hash:
            self.state.data_hash = data_hash
            self.on.k8s_service_available.emit()
```

Commit:
Address Corys comments

Update Relation values during initialization
instead of using events.

The discussion for these changes are here:
https://github.com/charmed-kubernetes/interface-k8s-service/pull/1

Signed-off-by: Dominik Fleischmann <dominik.fleischmann@canonical.com>

johnsca commented on k8s_service.py:
```diff
@@ -0,0 +1,48 @@
+                self.services.append((service_name, service_port))
+
+        self.is_available = bool(self.services)
+        self.is_created = bool(self.model.relations[relation_name])
```
I'm having second thoughts on the name of this flag.  Just `is_created` feels a bit too ambiguous and almost sounds more like it's indicating whether the object instance has been created, though that's a bit nonsensical. Maybe `is_relation_created` instead, or shortened to `is_related`?  If we end up with a common base class for these types of interface API libraries, we might want it to automatically always provide `is_relation_created` and `is_relation_joined` (or `is_related` and `is_joined` if we think that's sufficient).  I know @kwmonroe will want to weigh in here, too, so that we avoid the joined / connected / available / ready fiasco from reactive. 😉

johnsca reviewed (approved):
I think this is good to land, with the understanding that we'll continue to iterate on it based on evolving ideas and feedback / discussions from the Charm Craft team.

johnsca commented:
Items to discuss with the Charm Craft team:
* How do we strongly encourage (or require) interface API authors to use versioned API protocols from the start?
* Where should interface API components live / how should they be published?
* Does this approach seem like the best pattern to encourage; specifically, of eschewing event handling in interface API components and instead using `__init__` to examine the state of the model and distill it down to a more semantically relevant view?
  * Pros:
    * Code is very linear and easy to follow
    * Avoids race conditions and confusion around when hooks fire and what events need to be observed
  * Cons:
    * Obscures the nature of the relation hooks, potentially leading to less understanding of how that process works
    * Leads to a catch-22 if components want to emit custom events for the charm, since it would have to be done during `__init__` before the charm could possibly have registered any observers

Regarding that last one, I can see a couple of potential options:
* The framework could collect any events "emitted" during the charm instance construction and hold them until between the "reemit deferred events" and "emit charm events" steps
* Components could support an `emit_events` method or similar which is invoked by the framework at the appropriate time
* Components could avoid creating any custom events at all and simply present the distilled view of the state of the model, then leave it up to the charm to translate that into custom events if so desired
